ledger/participant-state/kvutils/src/main/scala/com/daml/ledger/participant/state/kvutils/api/WriteServiceWithDeduplicationSupport.scala

```diff
@@ -0,0 +1,112 @@
+// Copyright (c) 2021 Digital Asset (Switzerland) GmbH and/or its affiliates. All rights reserved.
+// SPDX-License-Identifier: Apache-2.0
+
+package com.daml.ledger.participant.state.kvutils.api
+
+import java.util.concurrent.CompletionStage
+
+import akka.stream.Materializer
+import com.daml.daml_lf_dev.DamlLf
+import com.daml.error.DamlContextualizedErrorLogger
+import com.daml.ledger.api.domain.ApplicationId
+import com.daml.ledger.api.health.HealthStatus
+import com.daml.ledger.configuration.Configuration
+import com.daml.ledger.offset.Offset
+import com.daml.ledger.participant.state.kvutils.deduplication.DeduplicationPeriodSupport
+import com.daml.ledger.participant.state.v2.{
+  PruningResult,
+  SubmissionResult,
+  SubmitterInfo,
+  TransactionMeta,
+  WriteService,
+}
+import com.daml.lf.data.Ref.{Party, SubmissionId}
+import com.daml.lf.data.Time
+import com.daml.lf.transaction.SubmittedTransaction
+import com.daml.logging.{ContextualizedLogger, LoggingContext}
+import com.daml.telemetry.TelemetryContext
+
+import scala.concurrent.ExecutionContext
+import scala.jdk.FutureConverters._
+
+class WriteServiceWithDeduplicationSupport(
```

Comment:
I think it's key to note that this component supports deduplication *periods*, i.e., the existing `WriteService` implementations may already support deduplication. Alternative names to consider: `DeduplicationPeriodSupportingWriteService`, `DeduplicationPeriodSupportWriteServiceDecorator`, or simply `DeduplicationPeriodSupportDecorator`.